planner/core/cache.go

```diff
@@ -87,6 +87,19 @@ func (key *pstmtPlanCacheKey) Hash() []byte {
 	return key.hash
 }
 
+// PstmtCacheKeyMutator is the interface that change Prepared statement's cache key.
+type PstmtCacheKeyMutator interface {
+	SetPstmtIDSchemaVersion(pstmtID uint32, schemaVersion int64)
+}
+
+// SetPstmtIDSchemaVersion implements PstmtCacheKeyMutator interface to change pstmtID and schemaVersion of cacheKey.
+// so we can reuse Key instead of new every time.
+func (key *pstmtPlanCacheKey) SetPstmtIDSchemaVersion(pstmtID uint32, schemaVersion int64) {
+	key.pstmtID = pstmtID
+	key.schemaVersion = schemaVersion
+	key.hash = nil
```

Comment:
It's better to set `key.hash = key.hash[:0]`. In https://github.com/pingcap/tidb/pull/8332/files#diff-76a70a17b419c3333a3ff060d8f7c330R73:
```go
if len(key.hash) == 0 {
	// calculate hash value
}
```

Reply:
`key.hash = key.hash[:0]` will not release the memory and it's some kind of leak, while `key.hash = nil` does

Reply:
@tiancaiamao I reuse this key because it can be reuse if there have more then one key(emm, we can new everytime too), if we keep reuse way `[:0]` in here is more suitable.